lib/rules/component-options-name-casing.js

```diff
@@ -0,0 +1,77 @@
+/**
+ * @author Pig Fang
+ * See LICENSE file in root directory for full license.
+ */
+'use strict'
+
+// ------------------------------------------------------------------------------
+// Requirements
+// ------------------------------------------------------------------------------
+
+const utils = require('../utils')
+const casing = require('../utils/casing')
+
+// ------------------------------------------------------------------------------
+// Rule Definition
+// ------------------------------------------------------------------------------
+
+module.exports = {
+  meta: {
+    type: 'suggestion',
+    docs: {
+      description:
+        'enforce the casing of component name in `components` options',
+      categories: undefined,
+      url: 'https://eslint.vuejs.org/rules/component-options-name-casing.html'
+    },
+    fixable: 'code',
+    schema: [{ enum: casing.allowedCaseOptions }],
+    messages: {
+      caseNotMatched: 'Component name "{{component}}" is not {{caseType}}.'
+    }
+  },
+  /** @param {RuleContext} context */
+  create(context) {
+    const caseType = context.options[0] || 'PascalCase'
+
+    const checkCase = casing.getChecker(caseType)
+    const convert = casing.getConverter(caseType)
+
+    return utils.executeOnVue(context, (obj) => {
+      const node = utils.findProperty(obj, 'components')
+      if (!node || node.value.type !== 'ObjectExpression') {
+        return
+      }
+
+      node.value.properties.forEach((property) => {
+        if (
+          property.type === 'Property' &&
+          property.key.type === 'Identifier'
+        ) {
+          const { name } = property.key
+          if (checkCase(name)) {
+            return
+          }
+
+          context.report({
+            node: property.key,
+            messageId: 'caseNotMatched',
+            data: {
+              component: name,
+              caseType
+            },
+            fix: (fixer) => {
+              const converted = convert(name)
+              if (caseType === 'kebab-case') {
+                return property.shorthand
+                  ? fixer.replaceText(property, `'${converted}': ${name}`)
+                  : fixer.replaceText(property.key, `'${converted}'`)
+              }
+              return fixer.replaceText(property.key, converted)
```

Comment:
Could you change it so that only cases that can be safely converted are auto-fixed?

I think the conversion to PascalCase is safe. And I think the kebab-case component can be converted to camelCase.
However, if it was used in camelCase or PascalCase in the template, it will be broken and cannot be converted to kebab-case.


You can check it on [Vue SFC Playground].

[Vue SFC Playground]: https://sfc.vuejs.org/#eyJBcHAudnVlIjoiPHNjcmlwdD5cbmltcG9ydCBNeUJ1dHRvbiBmcm9tICcuL015QnV0dG9uLnZ1ZSdcbmV4cG9ydCBkZWZhdWx0IHtcblx0Y29tcG9uZW50czoge1xuICAgIFBhc2NhbENhc2U6IE15QnV0dG9uLFxuICAgICdrZWJhYi1jYXNlJzogTXlCdXR0b24sXG4gICAgY2FtZWxDYXNlOiBNeUJ1dHRvbixcbiAgfVxufVxuPC9zY3JpcHQ+XG5cbjx0ZW1wbGF0ZT5cbiAgPHBhc2NhbC1jYXNlPnBhc2NhbC1jYXNlPC9wYXNjYWwtY2FzZT58XG4gIDxQYXNjYWxDYXNlPlBhc2NhbENhc2U8L1Bhc2NhbENhc2U+fFxuICA8cGFzY2FsQ2FzZT5wYXNjYWxDYXNlPC9wYXNjYWxDYXNlPlxuICA8YnI+XG4gIDxrZWJhYi1jYXNlPmtlYmFiLWNhc2U8L2tlYmFiLWNhc2U+fFxuICA8S2ViYWJDYXNlPktlYmFiQ2FzZTwvS2ViYWJDYXNlPnxcbiAgPGtlYmFiQ2FzZT5rZWJhYkNhc2U8L2tlYmFiQ2FzZT5cbiAgPGJyPlxuICA8Y2FtZWwtY2FzZT5jYW1lbC1jYXNlPC9jYW1lbC1jYXNlPnxcbiAgPENhbWVsQ2FzZT5DYW1lbENhc2U8L0NhbWVsQ2FzZT58XG4gIDxjYW1lbENhc2U+Y2FtZWxDYXNlPC9jYW1lbENhc2U+XG48L3RlbXBsYXRlPiIsImltcG9ydC1tYXAuanNvbiI6IntcbiAgXCJpbXBvcnRzXCI6IHtcbiAgICBcInZ1ZVwiOiBcImh0dHBzOi8vc2ZjLnZ1ZWpzLm9yZy92dWUucnVudGltZS5lc20tYnJvd3Nlci5qc1wiXG4gIH1cbn0iLCJNeUJ1dHRvbi52dWUiOiI8dGVtcGxhdGU+XG4gIDxidXR0b24+PHNsb3Q+TXlCdXR0b248L3Nsb3Q+PC9idXR0b24+XG48L3RlbXBsYXRlPiJ9

Reply:
Unsafe cases could use ESLint's suggestion API (which is basically a fixer that is only called manually by the user). Then you'd give a starting point for refactorings. Also please mention this case in the docs.